File: src/main/java/com/github/pulsebeat02/test/YoutubeResourcepackTest.java
package com.github.pulsebeat02.test;

import com.github.pulsebeat02.MinecraftMediaLibrary;
import com.github.pulsebeat02.concurrent.AsyncVideoExtraction;
import com.github.pulsebeat02.extractor.YoutubeExtraction;
import com.github.pulsebeat02.image.ImageMap;
import com.github.pulsebeat02.resourcepack.ResourcepackWrapper;
import com.github.pulsebeat02.resourcepack.hosting.HttpDaemonProvider;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class YoutubeResourcepackTest extends JavaPlugin {

    private MinecraftMediaLibrary library;

    @Override
    public void onEnable() {
        library = new MinecraftMediaLibrary(this, getDataFolder().getPath(), true);
    }

    public String getResourcepackUrlYoutube(@NotNull final String youtubeUrl, @NotNull final String directory, final int port) {

        YoutubeExtraction extraction = new YoutubeExtraction(youtubeUrl, directory) {
            @Override
            public void onVideoDownload() {
                System.out.println("Video is Downloading!");
            }

            @Override
            public void onAudioExtraction() {
                System.out.println("Audio is being extracted from Video!");
            }
        };
        ExecutorService executor = Executors.newCachedThreadPool();
        CompletableFuture.runAsync(() -> new AsyncVideoExtraction(extraction).extractAudio(), executor);
        CompletableFuture.runAsync(() -> new AsyncVideoExtraction(extraction).downloadVideo(), executor);

        ResourcepackWrapper wrapper = new ResourcepackWrapper.Builder()
                .setAudio(extraction.getAudio())
                .setDescription("Youtube Video: " + extraction.getVideoTitle())
                .setPath(directory)
                .setPackFormat(6)
                .createResourcepackHostingProvider(library);
        wrapper.buildResourcePack();

        HttpDaemonProvider hosting = new HttpDaemonProvider(directory, port);
        hosting.startServer();

        return hosting.generateUrl(Paths.get(directory));

    }

    public void displayImage(final int map, @NotNull final File image) throws IOException {

        BufferedImage bi = ImageIO.read(image);

        ImageMap imageMap = new ImageMap.Builder()
                .setMap(map)
                .setViewers(getOnlinePlayerUUIDs())
                .setWidth(bi.getWidth())
                .setHeight(bi.getHeight())
                .createImageMap(library);

        imageMap.drawImage();

    }

    public UUID[] getOnlinePlayerUUIDs() {
        List<? extends Player> players = new ArrayList<>(Bukkit.getOnlinePlayers());
        UUID[] uuids = new UUID[players.size()];
        for (int i = 0; i < players.size(); i++) {
            uuids[i] = players.get(i).getUniqueId();
        }
        return uuids;
    }

}

File: MinecraftMediaLibrary-API/src/main/test/com/github/pulsebeat02/minecraftmedialibrary/test/video/YoutubeResourcepackTest.kt
/*
 * ============================================================================
 * Copyright (C) PulseBeat_02 - All Rights Reserved
 *
 * This file is part of MinecraftMediaLibrary
 *
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 *
 * Written by Brandon Li <brandonli2006ma@gmail.com>, 2/26/2021
 * ============================================================================
 */
package com.github.pulsebeat02.minecraftmedialibrary.test.video

import com.github.pulsebeat02.minecraftmedialibrary.MinecraftMediaLibrary
import com.github.pulsebeat02.minecraftmedialibrary.concurrent.AsyncVideoExtraction
import com.github.pulsebeat02.minecraftmedialibrary.extractor.ExtractionSetting
import com.github.pulsebeat02.minecraftmedialibrary.extractor.YoutubeExtraction
import com.github.pulsebeat02.minecraftmedialibrary.image.basic.MinecraftStaticImage
import com.github.pulsebeat02.minecraftmedialibrary.resourcepack.ResourcepackWrapper
import com.github.pulsebeat02.minecraftmedialibrary.resourcepack.hosting.HttpDaemonProvider
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.io.IOException
import java.nio.file.Paths
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors
import javax.imageio.ImageIO

class YoutubeResourcepackTest : JavaPlugin() {

    private var library: MinecraftMediaLibrary? = null

    override fun onEnable() {
        library = MinecraftMediaLibrary(this)
    }

    fun getResourcepackUrlYoutube(youtubeUrl: String, directory: String, port: Int): String {
        val extraction: YoutubeExtraction = object : YoutubeExtraction(
            youtubeUrl, directory, ExtractionSetting.builder().build()
        ) {
            override fun onVideoDownload() {
                println("Video is Downloading!")
            }

            override fun onAudioExtraction() {
                println("Audio is being extracted from Video!")
            }
        }
        val executor = Executors.newCachedThreadPool()
        CompletableFuture.runAsync({ AsyncVideoExtraction(extraction).extractAudio() }, executor)
        CompletableFuture.runAsync(
            { AsyncVideoExtraction(extraction).downloadVideo() }, executor
        )
        val wrapper = ResourcepackWrapper.builder()
            .setAudio(extraction.audio)
            .setDescription("Youtube Video: " + extraction.videoTitle)
            .setPath(directory)
            .setPackFormat(6)
            .build(library)
        wrapper.buildResourcePack()
        val hosting = HttpDaemonProvider(directory, port)
        hosting.startServer()
        return hosting.generateUrl(Paths.get(directory))
    }

    @Throws(IOException::class)
    fun displayImage(map: Int, image: File) {
        val bi = ImageIO.read(image)
        val imageMap = MinecraftStaticImage.builder()
            .setMap(map)
            .setWidth(bi.width)
            .setHeight(bi.height)
            .build(library)
        imageMap.drawImage()
    }
}